### ISSUE TITLE
Shorter way to render LaTeX

### ISSUE THREAD

**Issue by lluni** [silverbulletmd/silverbullet-katex#1]
In for example Logseq or Obsidian it is possible to surround a given text with `$`/`$$` to render the text inline/multi-line (kind of like the current option with the LaTeX code block, but usually centered like `align*`/`equation*` if not specified otherwise) respectively.

This would make it faster and simpler to write LaTeX as there wouldn't be as much overhead to make a code block anymore.

**Comment by showgood163** [comment]
Agree. Inline math formatting and rendering support are very important.

**Comment by showgood163** [comment]
Will this issue ever be updated?

**Comment by alvorithm** [comment]
Hi @zefhemel -- I think this and #4 are essentially asking for inline rendering of math, which in my experience is going to be needed by anybody who needs math. 

In other words: I could be wrong, but a plugin that only does display (=paragraph) math is not useful for the vast majority of the math-using audience. 

The architectural question is: is it possible to extend inline styles with the plugin mechanism, or does it require deeper changes in silverbullet?

Thank you so much for bringing silverbullet to the world!

**Comment by zefhemel** [comment]
Yes, I agree and this does indeed make things significantly harder. This cannot be done with a plug and requires extending the existing markdown parser built into SilverBullet.

**Comment by alvorithm** [comment]
I had a quick search. Since I am unfamiliar with all of Lezer, TS and KaTeX/Mathjax, I really hope these hits are relevant and help in to make math typesetting fully supported in silverbullet: it is such an central use-case for many students, scientists and engineers!

* [Replacing text between dollar signs...](https://discuss.codemirror.net/t/replacing-text-between-dollar-signs-for-mathml-expression/3850) (somewhat like the strikethrough extension).
* [Identifying `$ $` and `$$ $$`-delimited strings for code highlighting](https://discuss.codemirror.net/t/markdown-and-latex-syntax-highlighting/4382/11), with [code](https://github.com/laurent22/joplin/blob/689931605722ca6d57cafe06b15089613e641b8b/packages/app-mobile/components/NoteEditor/MarkdownTeXParser.ts)
* [Jupyterlab's handling of MathJax](https://github.com/jupyterlab/jupyterlab/blob/f14c89145d697974ecbd02872643b530381d90b6/packages/mathjax-extension/src/index.ts)

**Comment by zefhemel** [comment]
Yes those seem relevant. What makes this harder though is that the `$bla` syntax is already in use for anchors: https://silverbullet.md/Markdown/Anchors so somehow the parser would have to be able to disambiguate between `$someFormula$` being parsed as a LaTeX formula and an anchor ending with a `$`. Perhaps forcing to put spaces around would be acceptable (`$ someFormula $`)?

**Comment by alvorithm** [comment]
Yes, that is unfortunate. 

If it were up to my sole personal preference (others should chip in!) I'd rather disallow anchors ending with `$` (this seems not too unreasonable given the semantics of `$` as currency unit?)

Typing formulas is already quite cumbersome, so it'd be nice if it could be as efficient as possible.
([Asciimath](http://asciimath.org/) is a thing, but I think there's no way around the more verbose LaTeX syntax in terms of mindshare.)

For display math, it would be great if the conventional `$$ ... $$` (with or without extra spaces, as you may decide) could be adopted. It rolls somewhat better off the fingers than the current latex block, which could be the syntax for syntax-highlighted (La)TeX code.

What do you think?

**Comment by alvorithm** [comment]
This discussion reminded me of the extensible markdown language designed for Pharo Smalltalk, [Microdown (pdf paper)](https://rmod-files.lille.inria.fr/Team/Texts/Papers/Duca20a-Microdown-IWST.pdf). One of their highlights is that they provide for labeled equations, i.e. equations with anchors. Considering this scenario where the two syntaxes interact might help find a way forward.

Their approach to anchors makes it possible to link to figures, and to parameterize code. One of their design goals was for Microdown to remain a subset of Markdown. 

Concretely, this how they do [math support in microdown](https://github.com/pillar-markup/Microdown?tab=readme-ov-file#math-support).

Something to highlight is that their `@anchor` is a block element, i.e. must stand on its own line, unlike `$anchor` in silverbullet.

**Comment by alvorithm** [comment]
Something I had not considered above is how `$` doubling down for anchors and the start of inline math impacts the editing experience. I can think of two aspects to consider:

* `$f` will start rendering greyed out as an anchor and then be interpreted as a formula upon closing of `$f` with `(x)=x^2$`. This might be slightly unexpected.
* not having a unique identifiable start token for math makes progressive rendering impossible even if it one day has good library support (I mean when typing `$\sum_{i} ...` the Σ renders upon typing `_` then the subindex on closing with `}`, etc. )

**Comment by abhemanyus** [comment]
Just chipping in, could an inline code block work, like pandoc's \`x = 4\`{.ruby}?

**Comment by v-shenoy** [comment]
Still kind of new to SilverBullet so forgive me for lack of knowledge. Could you elaborate a bit more on why this cannot be achieved via plugs?

**Comment by v-shenoy** [comment]
I would not mind trying to take a stab at this, considering this is one of the few critical features missing in my opinion. However, it would take me sometime to understand the internal workings of SilverBullet and its interactions with plugs. @zefhemel (also taking this opportunity to thank you for this cool project!)

**Comment by zefhemel** [comment]
So I'm struggling with this one, because in my mind this is a bit of a niche use case (using LaTeX formulas). It's not (currently) possible to extend the markdown parser (which lives here: https://github.com/silverbulletmd/silverbullet/tree/main/common/markdown_parser) via a plug, which means that this syntax extension (special handling of `$...$`) would need to be built into the markdown parser. Generally I'm not a huge fan of this, because I prefer generic solutions that solve more than one use case (like LaTeX formulas in this case).

One thing I'm working on in the context of [Space Lua](https://silverbullet.md/Space%20Lua) is to introduce another piece of non-standard Markdown syntax, namely `${Lua expression}`. Right now, if the result of this Lua expression returns a table (object) with a `html` key, that HTML will be rendered inline (see the marquee example on that linked page). This means that you could hypothetically build a Lua function `formula` (or whatever) that you can call like `${formula[[my fancy equation here]]}` (this is one of the syntaxes Lua allows to call a `formula` function with `my fancy equation` as a string argument) and let it be rendered via LaTeX somehow. 

Would that be a reasonable compromise? Still not as concise as simply `$my fancy formula$` but better than the fenced code block syntax.

**Comment by v-shenoy** [comment]
Syntactically, I think that is reasonable given that the Space Lua expressions aren't too inconvenient to write. However, I am a bit unsure on how one would go about writing the `formula` function, and the effort it would take to support mathematical symbols which are supported in LaTeX. 

Would it be possible to have this method part of the SilverBullet library then?

**Comment by alvorithm** [comment]
Space Lua sounds exciting! 

My .02€ is that a syntax involving the sigils `${[[]]}` is too heavy to use in practice. 

* if you want to LaTeX-render something quick like `$f(x)=x^2$` this syntax is a lot of overhead: 12 characters in the example invocation with a function named _formula_ (7 involving a modifier) vs 2 characters (2 involving a modifier) in order to type 8 characters.
* if you have a more involved expression, you'll also have a readability problem, since LaTeX expressions are bound to have braces, brackets, parentheses, and other sigils on top of the suggested invocation syntax. It will be painful to read and edit.

I can only speak for myself of course, but I also speculate that other heavy users of math are going to find this compromise too inconvenient.

**Comment by phxyz12** [comment]
Hi all

Just came accross this and wonder if that topic is still active. If not, does this mean that the Latex support can be considered to be more or less dropped, at least for the time being?

Please don't get me wrong, I appreciate the whole project and all the efforts of those who contribute with features very much. I am asking this since it doesn't make sense for me to start migrating a large notes collection that contain also lots of math if it is doubtful if there will be a solution.

Regarding the discussion in this thread, a few thoughs from me:

It was stated that LaTeX formulae are a 'niche use case'. This may be true when one removey all students and professionals from the natural sciences, economic sciences, computer sciences, engineering and a buch of others from the intended user base. LaTeX is the defacto standard for math type setting since the early 90ies of the last century. That's the reason why so many Markdown cappable products support LaTeX formula setting. I would certainly not consider this to be a niche. As long as you don't study old languages, you likely will require to write some math :-)

**Comment by bgamari** [comment]
I, too, have considered migrating my own notes, currently in Hedgedoc, to Silverbullet. However, the lack of inline math is rather a dealbreaker. The suggested syntax is indeed far too noisy.

As @phxyz12 suggests, many (perhaps even most?) potential users of Silverbullet rely heavily on mathematical notation; it seems odd that this should be considered to be a niche use-case.

**Comment by MrMugame** [comment]
This [library](https://github.com/MrMugame/silverbullet-math) (Disclaimer: My own) implements such a `formula` function. It's about as good as you can currently get (apart from shortening the name, which doesn't have a lot of benefit imo). Together with slash commands, it's definitely usable, but far from optimal, especially considering compatability.

<img width="767" height="432" alt="Image" src="https://github.com/user-attachments/assets/97d31c82-2b49-43da-907a-3dbafb51ddd4" />

**Comment by zefhemel** [comment]
I will be archiving this repository and point to @MrMugame's library instead. It is the better and more modern approach.
